File: julia/7_tebd.jl
#!/usr/bin/env julia
# -*- coding: utf-8 -*-
"""
(Simplified) TEBD simulation of random quantum circuits — Julia
"""

using LinearAlgebra
using Random
using TensorOperations

const C64 = ComplexF64

"単一量子ビットのランダムU(α,θ,φ)"
function random_u()::Matrix{C64}
    α   = 2π * rand()
    θ   = π  * rand()
    ϕ   = 2π * rand()
    eα  = cis(α)
    e₊ϕ = cis(ϕ)
    e₋ϕ = cis(-ϕ)
    return eα * [
        cos(θ/2)            -e₊ϕ*sin(θ/2)
        e₋ϕ*sin(θ/2)         cos(θ/2)
    ]
end

"state(2,2,...,2) へ pos の1量子ビットUを適用（全状態テンソル版）"
function apply_1q!(state::Array{C64}, pos::Int, U::Matrix{C64})
    n = ndims(state)
    order = vcat(pos, setdiff(1:n, pos))
    invorder = invperm(order)
    st = permutedims(state, order)                # (2, 2, ..., 2)
    st2 = reshape(st, 2, :)
    st2 = U * st2                                  # 左から作用
    st = reshape(st2, size(st))
    state .= permutedims(st, invorder)
    return state
end

"state(2,2,...,2) へ pos,pos+1 の2量子ビットG(2,2,2,2) を適用"
function apply_2q!(state::Array{C64}, pos::Int, G::Array{C64,4})
    n = ndims(state)
    order = vcat(pos, pos+1, setdiff(1:n, [pos, pos+1]))
    invorder = invperm(order)
    st = permutedims(state, order)                 # (2,2,2,2, ..., 2)
    st2 = reshape(st, 2, 2, :)
    @tensor st3[a,b,t] := G[a,b,c,d] * st2[c,d,t] # 先頭2脚へゲート
    st = reshape(st3, size(st))
    state .= permutedims(st, invorder)
    return state
end

"2サイトMPS更新: mps[i], mps[i+1] と G(2,2,2,2) からSVD分割（cutoff/χmax）"
function mps_two_site_update!(mps::Vector{Array{C64,3}}, i::Int;
                              G::Array{C64,4}, cutoff=1e-10, χmax=typemax(Int))
    A = mps[i]         # (χL, 2, r)
    B = mps[i+1]       # (r,  2, χR)

    # 2サイトにゲートを作用
    @tensor T[χl,i1,i2,χr] := G[i1,i2,j1,j2] * A[χl,j1,α] * B[α,j2,χr]
    Tmat = reshape(T, size(T,1)*size(T,2), size(T,3)*size(T,4))

    U,S,Vt = svd(Tmat; full=false)
    # しきい値/最大ボンド次元でトリム
    keep = findall(x -> x > cutoff, S)
    r = min(length(keep), χmax)
    r == 0 && (return (mps[i] .= zeros(C64, size(A,1),2,1);
                       mps[i+1] .= zeros(C64, 1,2,size(B,3))))
    S = S[1:r]
    U = U[:, 1:r]
    Vt = Vt[1:r, :]

    Ssqrt = Diagonal(sqrt.(S))
    Anew = U * Ssqrt
    Bnew = Ssqrt * Vt

    mps[i]   = reshape(Anew, size(A,1), 2, r)
    mps[i+1] = reshape(Bnew, r, 2, size(B,3))
end

function main()
    n       = 16      # qubits
    depth   = 16      # layers
    max_dim = 4       # χmax for truncated MPS
    cutoff  = 1e-10

    # CNOT (control = first, target = second)
    cnot = Array{C64}(undef, 2,2,2,2)
    cnot .= 0
    cnot[1,1,1,1] = 1
    cnot[1,2,1,2] = 1
    cnot[2,1,2,2] = 1
    cnot[2,2,2,1] = 1

    # |0...0> statevector (tensor形: (2,2,...,2))
    state = zeros(C64, ntuple(_->2, n))
    state[ntuple(_->1, n)...] = 1

    # MPS 初期化（|0⟩ = [1,0]^T）
    mps0 = [reshape([C64(1),C64(0)], 1,2,1) for _ in 1:n]
    mps1 = [copy(x) for x in mps0]
    println("mps0/1 initial virtual bond dimensions: ",
            [size(mps0[i],3) for i in 1:n-1])

    for k in 0:depth-1
        # ランダム1量子ゲート
        for pos in 1:n
            U = random_u()
            apply_1q!(state, pos, U)
            @tensor mps0[pos][a,i,b] := U[i,j] * mps0[pos][a,j,b]
            @tensor mps1[pos][a,i,b] := U[i,j] * mps1[pos][a,j,b]
        end

        # CNOT レイヤ（オッド/イーブン交互）
        for i in (k÷2 + 1):2:(n-1)
            apply_2q!(state, i, cnot)
            # 無切断
            mps_two_site_update!(mps0, i; G=cnot, cutoff=cutoff, χmax=typemax(Int))
            # 切断あり
            mps_two_site_update!(mps1, i; G=cnot, cutoff=cutoff, χmax=max_dim)
        end

        # MPS から状態再構成（mps0）
        ψ0 = mps0[1]
        ψ0 = reshape(ψ0, 2, size(ψ0,3))
        for i in 2:n
            @tensor tmp[a,c,d] := ψ0[a,b] * mps0[i][b,c,d]
            ψ0 = reshape(tmp, size(tmp,1)*size(tmp,2), size(tmp,3))
        end
        ψ0_vec = vec(permutedims(ψ0))  # row-major flatten

        # MPS から状態再構成（mps1）
        ψ1 = mps1[1]
        ψ1 = reshape(ψ1, 2, size(ψ1,3))
        for i in 2:n
            @tensor tmp[a,c,d] := ψ1[a,b] * mps1[i][b,c,d]
            ψ1 = reshape(tmp, size(tmp,1)*size(tmp,2), size(tmp,3))
        end
        ψ1_vec = vec(permutedims(ψ1))

        # フル state をベクトル化（row-major）
        st_mat = reshape(state, 2, :)
        st_vec = vec(permutedims(st_mat))

        fid0 = abs(dot(conj.(st_vec), ψ0_vec))^2
        fid1 = abs(dot(conj.(st_vec), ψ1_vec))^2
        println("step: $k fidelity: $fid0, $fid1")
    end

    println("mps0 final virtual bond dimensions: ", [size(mps0[i],3) for i in 1:n-1])
    println("mps1 final virtual bond dimensions: ", [size(mps1[i],3) for i in 1:n-1])
end

main()

File: rust/src/bin/7_tebd.rs
use anyhow::Result;
use ndarray::{Array1, Array2, Array3, Array4, array, s};
use ndarray_einsum::einsum;
use num_complex::Complex64 as C64;
use rand::distributions::{Distribution, Uniform};
use tn_basics::{EasySVD, MapStrToAnyhowErr};

fn iexp(theta: f64) -> C64 {
    C64::new(theta.cos(), theta.sin())
}

// (Simplified) TEBD simulation of random quantum circuits

fn main() -> Result<()> {
    let n: usize = 16;
    let depth: usize = 16;
    let max_dim: usize = 4;
    let cutoff: f64 = 1e-10;

    // two-qubit gate: CNOT
    let cnot: Array4<C64> = array![
        [1.0, 0.0, 0.0, 0.0],
        [0.0, 1.0, 0.0, 0.0],
        [0.0, 0.0, 0.0, 1.0],
        [0.0, 0.0, 1.0, 0.0]
    ]
    .mapv(|x: f64| C64::from(x))
    .into_shape_with_order((2, 2, 2, 2))?;

    // state vector |0...0>
    let mut state = Array1::<C64>::zeros([1usize << n]);
    state[0] = 1.0.into();
    //let state = state.into_shape_with_order(vec![2; n])?;

    // MPS |0...0> without truncation
    let mut mps0: Vec<Array3<C64>> = (0..n)
        .map(|_| array![[[1.0.into()], [0.0.into()]]])
        .collect();
    print!("mps0/1 initial virtual bond dimensions: [",);
    for i in 0..n {
        print!("{}, ", mps0[i].shape()[2]);
    }
    println!("]\n");

    // MPS |0...0> with truncation
    let mut mps1 = mps0.clone();

    let mut rng = rand::thread_rng();
    let uniform_2pi = Uniform::new(0.0, 2.0 * std::f64::consts::PI);
    let uniform_pi = Uniform::new(0.0, std::f64::consts::PI);

    for k in 0..depth {
        // random single-qubit rotations
        for pos in 0..n {
            let alpha = uniform_2pi.sample(&mut rng);
            let theta = uniform_pi.sample(&mut rng);
            let phi = uniform_2pi.sample(&mut rng);

            let u = iexp(alpha)
                * array![
                    [
                        C64::from((theta / 2.0).cos()),
                        -iexp(phi) * C64::from((theta / 2.0).sin())
                    ],
                    [
                        iexp(-phi) * C64::from((theta / 2.0).sin()),
                        C64::from((theta / 2.0).cos())
                    ]
                ];

            {
                // due to ndarray_einsum limitation, we apply gate by coonverting vector into fixed shape tensor
                let state_3d = state.to_shape((1usize << pos, 2, 1usize << (n - pos - 1)))?;
                let state_3d = einsum("ij,kjm->kim", &[&u, &state_3d]).map_str_err()?;
                state = state_3d.flatten().into_owned();
            }
            {
                let sub0 = einsum("ij,kjm->kim", &[&u, &mps0[pos]]).map_str_err()?;
                let [s0, s1, s2] = *sub0.shape() else {
                    panic!("unexpected shape")
                };
                mps0[pos] = sub0.into_shape_clone((s0, s1, s2))?;
            }
            {
                let sub1 = einsum("ij,kjm->kim", &[&u, &mps1[pos]]).map_str_err()?;
                let [s0, s1, s2] = *sub1.shape() else {
                    panic!("unexpected shape")
                };
                mps1[pos] = sub1.into_shape_clone((s0, s1, s2))?;
            }
        }

        // cnot gates
        for i in ((k / 2)..(n - 1)).step_by(2) {
            {
                // due to ndarray_einsum limitation, we apply gate by coonverting vector into fixed shape tensor
                let state_4d = state.to_shape((1usize << i, 2, 2, 1usize << (n - i - 2)))?;
                let state_4d = einsum("ijkl,mkln->mijn", &[&cnot, &state_4d]).map_str_err()?;
                state = state_4d.flatten().into_owned();
            }
            {
                let sub0 = einsum("ijkl,mkn->mijln", &[&cnot, &mps0[i]]).map_str_err()?;
                let sub0 = einsum("ijklm,mln->ijkn", &[&sub0, &mps0[i + 1]]).map_str_err()?;
                let [s0, s1, s2, s3] = *sub0.shape() else {
                    panic!("unexpected shape")
                };
                let tmat = sub0.to_shape((s0 * s1, s2 * s3))?;
                let (u, s, vt) = tmat.thin_svd()?;
                let rank_new = s
                    .iter()
                    .position(|&x| x <= cutoff * s[0])
                    .unwrap_or(s.len());
                let s = s.slice(s![0..rank_new]);
                let smat = Array2::from_diag(&s.mapv(|x| C64::from(x.sqrt())));
                mps0[i] = (u.slice_move(s![.., 0..rank_new]))
                    .dot(&smat)
                    .into_shape_clone((s0, s1, rank_new))?;
                mps0[i + 1] = smat
                    .dot(&(vt.slice_move(s![0..rank_new, ..])))
                    .into_shape_clone((rank_new, s2, s3))?;
            }
            {
                let sub1 = einsum("ijkl,mkn->mijln", &[&cnot, &mps1[i]]).map_str_err()?;
                let sub1 = einsum("ijklm,mln->ijkn", &[&sub1, &mps1[i + 1]]).map_str_err()?;
                let [s0, s1, s2, s3] = *sub1.shape() else {
                    panic!("unexpected shape")
                };
                let tmat = sub1.to_shape((s0 * s1, s2 * s3))?;
                let (u, s, vt) = tmat.thin_svd()?;
                let rank_new = s
                    .iter()
                    .position(|&x| x <= cutoff * s[0])
                    .unwrap_or(s.len())
                    .min(max_dim);
                let s = s.slice(s![0..rank_new]);
                let smat = Array2::from_diag(&s.mapv(|x| C64::from(x.sqrt())));
                mps1[i] = (u.slice_move(s![.., 0..rank_new]))
                    .dot(&smat)
                    .into_shape_clone((s0, s1, rank_new))?;
                mps1[i + 1] = smat
                    .dot(&(vt.slice_move(s![0..rank_new, ..])))
                    .into_shape_clone((rank_new, s2, s3))?;
            }
        }

        let (s0, s1, s2) = mps0[0].dim();
        let mut state0 = mps0[0]
            .clone()
            .into_shape_with_order((s0 * s1, s2))?
            .to_owned();
        for i in 1..n {
            let state = einsum("ij,jkl->ikl", &[&state0, &mps0[i]]).unwrap();
            let shape = (state.shape()[0] * state.shape()[1], state.shape()[2]);
            state0 = state.into_shape_clone(shape)?;
        }
        let state0 = state0.flatten();

        let (s0, s1, s2) = mps1[0].dim();
        let mut state1 = mps1[0]
            .clone()
            .into_shape_with_order((s0 * s1, s2))?
            .to_owned();
        for i in 1..n {
            let state = einsum("ij,jkl->ikl", &[&state1, &mps1[i]]).unwrap();
            let shape = (state.shape()[0] * state.shape()[1], state.shape()[2]);
            state1 = state.into_shape_clone(shape)?;
        }
        let state1 = state1.flatten();

        let state_conj = state0.mapv(|x| x.conj());

        println!(
            "step: {} fidelity: {}, {}",
            k,
            state_conj.dot(&state0).norm_sqr(),
            state_conj.dot(&state1).norm_sqr(),
        );
    }

    print!("mps0 final virtual bond dimensions: [",);
    for i in 0..(n - 1) {
        print!("{}, ", mps0[i].shape()[2]);
    }
    println!("]\n");
    print!("mps1 final virtual bond dimensions: [",);
    for i in 0..(n - 1) {
        print!("{}, ", mps1[i].shape()[2]);
    }
    println!("]\n");
    Ok(())
}
